### PR TITLE
Restore scroll when previous state was patch

### LINKED ISSUE #2080: live_redirect does not remember scroll position if a live_patch is done before it
### Environment

* Elixir version (elixir -v): 1.13.4
* Phoenix version (mix deps): 1.6.10
* Phoenix LiveView version (mix deps): 0.17.10
* Browsers you attempted to reproduce this bug on (the more the merrier):  chrome, firefox, safari
* Does the problem persist after removing "assets/node_modules" and trying again? Yes: yes

### Actual behavior

setup: liveview A and liveview B

SITUATION 1
1. load liveview A and scroll down
2. cilck on a live_redirect to liveview B
3. browser back returns to original scroll position of liveview A

SITUATION 2 (same as SITUATION 1 but with a live_patch in between)
1. load liveview A and scroll down
2. do live_patch, scroll more down 
3. cilck on a live_redirect to liveview B
4. browser back *does not* return to original scroll position of liveview A

### Expected behavior

I did not expect that a live_patch before a live_redirect would change the way the back button works w.r.t. the scroll position.

Is this intended behavior or a  possible bug?

Actual usecase: 

List view of items with a "load more button" (which is the live_patch), I'd like to be able to click on one item (liveview B) and go back to the original position in the list.

### PR DESCRIPTION
**Pull request by dvic**
Fixes #2080

### CODE REVIEW

**Comment by janwillemvd** [comment]
Nice, I can confirm this fix fixes the issue in two of our repos having this problem.

File: assets/js/phoenix_live_view/live_socket.js [hunk 1]
```diff
@@ -692,6 +690,14 @@ export default class LiveSocket {
     }, false)
   }
 
+  maybeScroll(scroll) {
+    if(typeof(scroll) === "number"){
+      setTimeout(() => {
+        window.scrollTo(0, scroll)
+      }, 0) // the body needs to render before we scroll.
```
Comment (davydog187) [inline assets/js/phoenix_live_view/live_socket.js:697]: This should probably use requestAnimationFrame instead of a timeout

Comment (dvic) [reply assets/js/phoenix_live_view/live_socket.js:697]: I was not sure if there was a reason why it was a timeout in the first place, so I left that unchanged. But I guess I can change it to requestAnimationFrame.

**Comment by jeregrine** [comment]
I wish this kinda thing were easier to test lol. This LGTM

**Comment by dvic** [comment]
> I wish this kinda thing were easier to test lol. This LGTM

If there's interest, I could contribute a PR that adds an integration test project with Wallaby set up? It seems like a good idea to be able to do fully fledged browser tests?

**Comment by chrismccord** [comment]
I don't believe this is what we want because live patch shouldn't change the scroll position. Since the issue is on the live redirect side, I believe this change is fixing it by accident? Live patch is about changing the URL on the current LV and shouldn't attempt to scroll to a new position.

**Comment by dvic** [comment]
> I don't believe this is what we want because live patch shouldn't change the scroll position. Since the issue is on the live redirect side, I believe this change is fixing it by accident? Live patch is about changing the URL on the current LV and shouldn't attempt to scroll to a new position.

Is there a particular example where this behaviour is unwanted? All the examples I can think of (pagination, applying filters and clicking on the item and going back) it's actually behaviour you expect as a user.

For example: when you click a few times on a "load more" button and you keep scrolling through the list, when you click the back button, you expect to go back to the position you previously clicked the pagination button right? (i.e., the bottom of the list before you clicked the button)

**Comment by dvic** [comment]
> Live patch is about changing the URL on the current LV and shouldn't attempt to scroll to a new position.

Just to add to the above: I agree with this sentence, but clicking on the back button *should* change the scroll and that's what this PR is doing: it's being applied in the `popstate` event.

**Comment by dvic** [comment]
In any case this escaped your attention (with all the incredible amount of work going on): is there anything I can clear up / check? I was really hoping to get this landed soon 😅

**Comment by jeregrine** [comment]
Thanks @dvic and @janwillemvd we're not in 100% full agreement on behavior here but we're going to accept this patch and roll with the punches. 

<3 Thank you so much for your patience and effort here.

**Pull request merged**